File: collectHeightWeight.py
#Write a program to collect height & weight of N students. Find the highest weight/height ratio using linked list.

class Student:
    def __init__(self, height, weight):
        self.height = height
        self.weight = weight
        self.ratio = weight / height
        self.next = None

def insert(head, height, weight):
    new_node = Student(height, weight)
    if head is None:
        return new_node
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new_node
    return head

def find_max_ratio(head):
    max_ratio = 0.0
    while head:
        if head.ratio > max_ratio:
            max_ratio = head.ratio
        head = head.next
    return max_ratio

# Input and execution
n = int(input("Enter number of students: "))
head = None
for i in range(n):
    h, w = map(float, input(f"Enter height and weight of student {i+1}: ").split())
    head = insert(head, h, w)

print("Highest weight/height ratio is: {:.2f}".format(find_max_ratio(head)))

File: collectHeightWeight.c
//Write a program to collect height & weight of N students. Find the highest weight/height ratio using linked list.

#include <stdio.h>
#include <stdlib.h>

// Define the Node structure
struct Student {
    float height;
    float weight;
    float ratio;
    struct Student* next;
};

// Function to create a new node
struct Student* createNode(float height, float weight) {
    struct Student* newNode = (struct Student*)malloc(sizeof(struct Student));
    newNode->height = height;
    newNode->weight = weight;
    newNode->ratio = weight / height;
    newNode->next = NULL;
    return newNode;
}

// Function to insert at the end of the list
void insert(struct Student** head, float height, float weight) {
    struct Student* newNode = createNode(height, weight);
    if (*head == NULL) {
        *head = newNode;
    } else {
        struct Student* temp = *head;
        while (temp->next != NULL)
            temp = temp->next;
        temp->next = newNode;
    }
}

// Function to find the maximum weight/height ratio
float findMaxRatio(struct Student* head) {
    float maxRatio = 0.0;
    while (head != NULL) {
        if (head->ratio > maxRatio)
            maxRatio = head->ratio;
        head = head->next;
    }
    return maxRatio;
}

int main() {
    struct Student* head = NULL;
    int n, i;
    float height, weight;

    printf("Enter the number of students: ");
    scanf("%d", &n);

    for (i = 0; i < n; i++) {
        printf("Enter height and weight of student %d: ", i + 1);
        scanf("%f %f", &height, &weight);
        insert(&head, height, weight);
    }

    float maxRatio = findMaxRatio(head);
    printf("Highest weight/height ratio is: %.2f\n", maxRatio);

    return 0;
}

